staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -89,30 +106,46 @@ const (
 type FeatureGate interface {
 	// Enabled returns true if the key is enabled.
 	Enabled(key Feature) bool
+	// FeatureSpec returns the FeatureSpec of the key exists as seen by the EmulationVersion, an error otherwise
+	FeatureSpec(key Feature) (FeatureSpec, error)
 	// KnownFeatures returns a slice of strings describing the FeatureGate's known features.
 	KnownFeatures() []string
 	// DeepCopy returns a deep copy of the FeatureGate object, such that gates can be
 	// set on the copy without mutating the original. This is useful for validating
 	// config against potential feature gate changes before committing those changes.
 	DeepCopy() MutableFeatureGate
+	EmulationVersion() *version.Version
 }
 
 // MutableFeatureGate parses and stores flag gates for known features from
 // a string like feature1=true,feature2=false,...
 type MutableFeatureGate interface {
 	FeatureGate
+	// SetEmulationVersion overrides the emulationVersion of the feature gate.
+	// Otherwise, the emulationVersion will be the same as the binary version.
+	// If set, the feature defaults and availability will be as if the binary is at the emulated version.
+	SetEmulationVersion(emulationVersion *version.Version) error
 
 	// AddFlag adds a flag for setting global feature gates to the specified FlagSet.
 	AddFlag(fs *pflag.FlagSet)
 	// Set parses and stores flag gates for known features
 	// from a string like feature1=true,feature2=false,...
+	// Set ignores SetFromMap error, because the emulationVersion may not be the final emulationVersion when the flag is set.
+	// Users should always call Validate() later to check for flag errors.
 	Set(value string) error
+	// Validate checks if the flag gates are valid at the emulated version.
+	// Should always be called after Set.
+	Validate() []error
```

Comment:
This change is to an exported shared library, but appears to be breaking. Do existing users users of MutableFeatureGate break at runtime if they don't realize to pick up this change? Should we introduce a new function and keep the existing behavior of `Set`?

Reply:
Existing user of this function is the flag parser, and it is not possible to make it call `Set` on some cases, and call `NewSet` on other cases. If we keep it the same, then the server would fail at the initial parsing stage if we ever want to set any feature that is not locked to default at the emulated version. I could not think of a way to keep the existing behavior of `Set` while supporting emulated version.

Reply:
The only user of this function is the flag parser, and we cannot change which function the parser calls, which has to be `Set`. If we keep its existing behavior, then the server would immediately fail whenever we try to set a feature flag which is `LockToDefault` at binary version but not emulated version. One workaround is to check if `value.contains("EmulationVersion")`, but that is hacky, and the library is not supposed to know any specific features.

Reply:
I see. Maybe offer a setting on the implementation of the interface that allows one to defer validation? It could default to disabled to preserve existing behavior and then enabled for our use case.

Reply:
> The only user of this function is the flag parser

The only in-tree user. Looks like there are many out-of-tree users: https://github.com/search?q=MutableFeatureGate&type=code